sdk/metric/aggregator/histogram/histogram.go

```diff
@@ -92,58 +116,82 @@ func New(desc *export.Descriptor, boundaries []core.Number) *Aggregator {
 func (c *Aggregator) Count() (int64, error) {
-	return int64(c.checkpoint.count.AsUint64()), nil
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return int64(c.checkpoint().count), nil
 }
 
 // Histogram returns the count of events in pre-determined buckets.
 func (c *Aggregator) Histogram() (aggregator.Buckets, error) {
-	return c.checkpoint.buckets, nil
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return c.checkpoint().buckets, nil
 }
 
 // Checkpoint saves the current state and resets the current state to
 // the empty set.  Since no locks are taken, there is a chance that
 // the independent Sum, Count and Bucket Count are not consistent with each
 // other.
 func (c *Aggregator) Checkpoint(ctx context.Context, desc *export.Descriptor) {
-	// N.B. There is no atomic operation that can update all three
-	// values at once without a memory allocation.
-	//
-	// This aggregator is intended to trade this correctness for
-	// speed.
-	//
-	// Therefore, atomically swap fields independently, knowing
-	// that individually the three parts of this aggregation could
-	// be spread across multiple collections in rare cases.
-
-	c.checkpoint.count.SetUint64(c.current.count.SwapUint64Atomic(0))
-	c.checkpoint.sum = c.current.sum.SwapNumberAtomic(core.Number(0))
-
-	for i := 0; i < len(c.checkpoint.buckets.Counts); i++ {
-		c.checkpoint.buckets.Counts[i].SetUint64(c.current.buckets.Counts[i].SwapUint64Atomic(0))
-	}
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	// clear the checkpoint state before making it the new current state.
+	c.resetCheckpoint()
+
+	// swap the current and checkpoint state by changing the least significant bit
+	// of the currentIdx
+	atomic.AddUint32(&c.currentIdx, 1)
+
+	// TODO(paivagustavo): there is a chance where the SDK calls Checkpoint and an update
+	//  to the new checkpoint is still occurring, could this be a problem?
+	//  Checkpoint is called for every active records and then it gets exported,
+	//  this reduces the chances of inconsistencies on the exporter.
+}
+
+// current returns the current state with the lower bit of currentIdx.
+func (c *Aggregator) current() *state {
+	return &c.states[c.currentIdx&1]
+}
+
+// checkpoint returns the checkpoint state by inverting the lower bit of currentIdx.
+func (c *Aggregator) checkpoint() *state {
+	return &c.states[^c.currentIdx&1]
+}
+
+func (c *Aggregator) resetCheckpoint() {
+	checkpoint := c.checkpoint()
+
+	checkpoint.count.SetUint64(0)
+	checkpoint.sum.SetNumber(core.Number(0))
+	checkpoint.buckets.Counts = make([]core.Number, len(checkpoint.buckets.Counts))
 }
 
 // Update adds the recorded measurement to the current data set.
 func (c *Aggregator) Update(_ context.Context, number core.Number, desc *export.Descriptor) error {
 	kind := desc.NumberKind()
 
-	c.current.count.AddUint64Atomic(1)
-	c.current.sum.AddNumberAtomic(kind, number)
+	current := c.current()
+	current.count.AddUint64Atomic(1)
+	current.sum.AddNumberAtomic(kind, number)
 
 	for i, boundary := range c.boundaries {
```

Comment:
Sort of off-topic, but I'd like to not see a linear search here. I think this would work:

```
  idx := sort.Search(len(c.boundaries), func(i int) bool {
    return number.CompareNumber(kind, boundary) >= 0
  })
  current.buckets.Counts[idx].AddUint64Atomic(1)
```

... but there's a memory allocation there, so it's not perfect. Linear search is probably faster than even a hand-coded-allocation-free binary search for a small number of buckets, it would take some benchmarking to know if this matters. 

Note that in the OpenMetrics exposition format, the Histogram type supports several different bucket configurations. There are "linear" buckets which use evenly spaced boundaries, there are "logarithmic" buckets, which use log-linear spacing between boundaries. If either of these special cases is used, we can do better than linear search or binary search typically using simple division (linear case) or computing a single logarithm + a division (log-linear case). These two special cases are probably more important than the arbitrary boundary case, which is when you begin looking at binary search.

So, I'm not sure what's best. A hand-coded-allocation-free binary search is still not the best outcome if the buckets are somehow linear. If the buckets are going to be linear or log-linear, it calls for new constructors to pass in a different sort of bucket-izer interface. If there's a "bucket configuration" interface that supports (a) listing boundaries and (b) performing the necessary lookup too, then this complexity can be isolated outside of the histogram aggregator itself.

Reply:
so a LinearBucket(min, max, number) returns you a bucket configuration for `number` evenly spaced buckets between `min` and `max` for example. the search function uses division.

when using absolute measures, in particular, a `LogarithmicBucket(minValue, growthFactor)` returns you a bucket configuration where the first bucket is `(-Inf, minValue)` and each value after that is  exponentially increasing by a factor of `growth`. the search function calculates a `math.Log()` to find the bucket, etc.

lastly there's a `ArbitraryBucket()` configuration that's like what we have now. it uses binary search or linear search, depending on how many buckets there are.

Reply:
All this are total valid considerations that we should address in the future, I think it is valid for us to create some issues to track this :)

Reply:
+1 on addressing and doing so in a future issue. I like the suggestion of addressing common bucket implementation.